src/databricks/labs/ucx/source_code/notebooks/sources.py

```diff
@@ -165,10 +166,22 @@ def __init__(self, ctx: LinterContext, path: Path, content: str | None = None):
         self._path: Path = path
         self._content = content
 
-    @cached_property
+    @property
     def _source_code(self) -> str:
-        encoding = locale.getpreferredencoding(False)
-        return self._path.read_text(encoding) if self._content is None else self._content
+        if self._content is None:
+            self._content = self._path.read_text(self._guess_encoding())
+        return self._content
+
+    def _guess_encoding(self):
+        path = self._path.as_posix()
+        count = min(32, os.path.getsize(path))
+        with open(path, 'rb') as _file:
+            raw = _file.read(count)
```

Comment:
```suggestion
        with self._path.open('rb') as _file:
            raw = _file.read(4)
```
(I think all BOMs are 4 bytes or less.)